### PR TITLE
registration

### PR DESCRIPTION
**Pull request by bodya4243**
1) Add User Entity, controller, repository, mapper, service, DTO.
2) Add Registration Exception

### CODE REVIEW

File: src/main/java/com/example/onlinebookstore/OnlineBookStoreApplication.java [hunk 1]
```diff
@@ -9,6 +10,8 @@
 public class OnlineBookStoreApplication {
     @Autowired
     private BookService bookService;
+    @Autowired
+    private UserService userService;
```
Comment (andrii-hoienko) [inline src/main/java/com/example/onlinebookstore/OnlineBookStoreApplication.java]: Better to use constructor injectrion

File: src/main/java/com/example/onlinebookstore/controller/AuthController.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+@RequiredArgsConstructor
+@RestController
+@RequestMapping("/auth")
+public class AuthController {
+    private final UserService userService;
+
+    @PostMapping("/registration")
+    @ResponseStatus(HttpStatus.CREATED)
+    public UserDto registerUser(@RequestBody UserRequestDto requestDto)
```
Comment (andrii-hoienko) [inline src/main/java/com/example/onlinebookstore/controller/AuthController.java]:
```suggestion
    public UserDto registerUser(@RequestBody @Valid UserRequestDto requestDto)
```

File: src/main/java/com/example/onlinebookstore/exception/RegistrationException.java [hunk 3]
```diff
@@ -0,0 +1,7 @@
+package com.example.onlinebookstore.exception;
+
+public class RegistrationException extends Exception {
```
Comment (andrii-hoienko) [inline src/main/java/com/example/onlinebookstore/exception/RegistrationException.java:3]: What is the reason to make this exception checked?

File: src/main/java/com/example/onlinebookstore/dto/UserRequestDto.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
+package com.example.onlinebookstore.dto;
+
+import jakarta.validation.constraints.Email;
+import jakarta.validation.constraints.NotNull;
+import lombok.Data;
+import org.hibernate.validator.constraints.Length;
+
+@Data
+public class UserRequestDto {
```
Comment (andrii-hoienko) [inline src/main/java/com/example/onlinebookstore/dto/UserRequestDto.java:13]: Where is logic to validate password and repeatPassword?

**Review by andrii-hoienko — CHANGES_REQUESTED** [review]
Good job, but some improvements required

File: src/main/java/com/example/onlinebookstore/model/User.java [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+import jakarta.persistence.Id;
+import jakarta.persistence.Table;
+import lombok.Getter;
+import lombok.Setter;
+
+@Entity
+@Getter
+@Setter
+@Table
+public class User {
```
Comment (andrii-hoienko) [inline src/main/java/com/example/onlinebookstore/model/User.java:20]: Please use soft delete

File: src/main/java/com/example/onlinebookstore/service/UserServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+@Service
+@RequiredArgsConstructor
+public class UserServiceImpl implements UserService {
+    private final UserRepository userRepository;
+    private final UserMapper userMapper;
+
+    @Override
+    public UserDto registerUser(UserRequestDto userRequestDto) {
+        if (userRepository.findByEmail(userRequestDto.getEmail()).isPresent()) {
+            throw new RegistrationException("Can't find register user");
```
Comment (andrii-hoienko) [inline src/main/java/com/example/onlinebookstore/service/UserServiceImpl.java:21]:
```suggestion
            throw new RegistrationException("User with this email: EMAIL already exists");
```

File: src/main/java/com/example/onlinebookstore/validation/FieldMatchValidator.java [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+import jakarta.validation.ConstraintValidatorContext;
+import java.util.Objects;
+
+public class FieldMatchValidator implements ConstraintValidator<FieldMatch, UserRequestDto> {
+
+    @Override
+    public boolean isValid(UserRequestDto userRequestDto,
+                           ConstraintValidatorContext constraintValidatorContext) {
+        return Objects.equals(userRequestDto.getPassword(), userRequestDto.getRepeatPassword());
+    }
```
Comment (andrii-hoienko) [inline src/main/java/com/example/onlinebookstore/validation/FieldMatchValidator.java]: Please make it work with 2 any fields, not only passwords

**Review by andrii-hoienko — CHANGES_REQUESTED** [review]
Good job, but some improvements required

File: src/main/java/com/example/onlinebookstore/exception/RegistrationException.java (hunk 3 above)
Comment (karma-o) [reply src/main/java/com/example/onlinebookstore/exception/RegistrationException.java:3]: Actually, this is specified by the requirements for this task:
> Add AuthenticationController with the following methods:
public UserResponseDto register(UserRegistrationRequestDto request) throws RegistrationException

File: src/main/java/com/example/onlinebookstore/exception/RegistrationException.java [hunk 8]
```diff
@@ -0,0 +1,7 @@
+package com.example.onlinebookstore.exception;
+
+public class RegistrationException extends RuntimeException {
```
Comment (karma-o) [inline src/main/java/com/example/onlinebookstore/exception/RegistrationException.java]: Actually task requirements imply that it should be a checked exception:
> Add AuthenticationController with the following methods:
public UserResponseDto register(UserRegistrationRequestDto request) throws RegistrationException

File: src/main/java/com/example/onlinebookstore/validation/FieldMatch.java [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+    Class<?>[] groups() default {};
+    Class<? extends Payload>[] payload() default {};
+    String first();
+    String second();
+
+    @Target({ElementType.TYPE, ElementType.ANNOTATION_TYPE})
+    @Retention(RetentionPolicy.RUNTIME)
+    @interface List {
+        FieldMatch[] value();
+    }
```
Comment (karma-o) [inline src/main/java/com/example/onlinebookstore/validation/FieldMatch.java:24]: not sure if we need that

File: src/main/java/com/example/onlinebookstore/validation/FieldMatch.java [hunk 10]
```diff
@@ -0,0 +1,25 @@
+package com.example.onlinebookstore.validation;
+
+import jakarta.validation.Constraint;
+import jakarta.validation.Payload;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+@Target({ElementType.TYPE, ElementType.ANNOTATION_TYPE})
```
Comment (karma-o) [inline src/main/java/com/example/onlinebookstore/validation/FieldMatch.java:10]: and I do not think we need `ANNOTATION_TYPE` as well

File: src/main/java/com/example/onlinebookstore/validation/FieldMatchValidator.java [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+        return Objects.equals(field1Value, field2Value);
+    }
+
+    private Object getFieldValue(UserRequestDto userRequestDto, String fieldName) {
+        try {
+            Field field = UserRequestDto.class.getDeclaredField(fieldName);
+            field.setAccessible(true);
+            return field.get(userRequestDto);
+        } catch (NoSuchFieldException | IllegalAccessException e) {
+            throw new RegistrationException("can't find the appropriate field");
```
Comment (karma-o) [inline src/main/java/com/example/onlinebookstore/validation/FieldMatchValidator.java]: * I wouldn't say that `RegistrationException` is the appropriate use here, we can use the same validator for user password update endpoint
* make ex message more informative if you have such an opportunity
```suggestion
            throw new IllegalArgumentException("can't find the appropriate field: " + fieldName);
```

File: src/main/resources/application.properties [hunk 12]
```diff
@@ -6,5 +6,5 @@ spring.datasource.password=${DB_PASSWORD}
 spring.datasource.driver-class-name=com.mysql.cj.jdbc.Driver
 server.port=8080
 server.servlet.context-path=/api
-spring.jpa.hibernate.ddl-auto=validate
+spring.jpa.hibernate.ddl-auto=create-drop
```
Comment (karma-o) [inline src/main/resources/application.properties]: why do you need that? use liquibase for creating your db schema and revert this change back to `validate`

File: src/main/java/com/example/onlinebookstore/model/User.java (hunk 5 above)
Comment (karma-o) [reply src/main/java/com/example/onlinebookstore/model/User.java:20]: There is no soft delete requirement for the user entity, and in general, we do not usually delete users, and we won't have a delete endpoint for it in this project. But I think you can still leave it as it is if you want

**Review by karma-o — APPROVED** [review]
Approved with several important comments, please take a look at them before merging!
